analyzers/src/SonarAnalyzer.CSharp/Rules/InvalidCastToInterfaceAnalyzer.cs

```diff
@@ -0,0 +1,101 @@
+    protected override bool EnableConcurrentExecution => false;
+
+    protected override void Initialize(SonarAnalysisContext context)
+    {
+        context.RegisterCompilationStartAction(
+            compilationStartContext =>
+            {
+                var interfaceImplementer = BuildTypeMap(compilationStartContext.Compilation.GlobalNamespace.GetAllNamedTypes());
+                compilationStartContext.RegisterNodeAction(
+                    c =>
+                    {
+                        var cast = (CastExpressionSyntax)c.Node;
+                        var interfaceType = c.SemanticModel.GetTypeInfo(cast.Type).Type as INamedTypeSymbol;
+                        var expressionType = c.SemanticModel.GetTypeInfo(cast.Expression).Type as INamedTypeSymbol;
+                        if (IsImpossibleCast(interfaceImplementer, interfaceType, expressionType))
+                        {
+                            var location = cast.Type.GetLocation();
+                            var interfaceTypeName = interfaceType.ToMinimalDisplayString(c.SemanticModel, location.SourceSpan.Start);
+                            var expressionTypeName = expressionType.ToMinimalDisplayString(c.SemanticModel, location.SourceSpan.Start);
+                            var message = expressionType.IsInterface() ? MessageInterface : MessageClass;
+                            c.ReportIssue(Diagnostic.Create(S1944, location, string.Format(message, expressionTypeName, interfaceTypeName)));
+                        }
+                    },
+                    SyntaxKind.CastExpression);
+            });
+    }
+
+    private static TypeMap BuildTypeMap(IEnumerable<INamedTypeSymbol> allTypes)
+    {
+        var ret = new TypeMap();
+        foreach (var type in allTypes)
+        {
+            if (type.IsInterface())
+            {
+                Add(type, type);
+            }
+            foreach (var @interface in type.AllInterfaces)
+            {
+                Add(@interface, type);
+            }
+        }
+        return ret;
+
+        void Add(INamedTypeSymbol key, INamedTypeSymbol value)
+        {
+            if (!ret.TryGetValue(key.OriginalDefinition, out var values))
+            {
+                values = new();
+                ret.Add(key.OriginalDefinition, values);
+            }
+            values.Add(value);
+        }
+    }
+
+    private static bool IsImpossibleCast(TypeMap interfaceImplementer, INamedTypeSymbol interfaceType, INamedTypeSymbol expressionType)
+    {
+        return interfaceType.IsInterface()
+            && ConcreteImplementationExists(interfaceType)
+            && expressionType is not null
+            && !expressionType.IsSealed
+            && !expressionType.Is(KnownType.System_Object)
+            && (!expressionType.IsInterface() || ConcreteImplementationExists(expressionType))
+            && interfaceImplementer.TryGetValue(interfaceType, out var implementers)
```

Comment:
```csharp
            && interfaceImplementer.TryGetValue(interfaceType, out var implementers)
```
Here the key given is `interfaceType` and above in `void Add(INamedTypeSymbol key, INamedTypeSymbol value)`
the key added is the `interfaceType.OriginalDefinition`. 

Won't this cause an issue? :/ 

PS: Also in line 47 where `IsImpossibleCast` is invoked I also see that `INamedTypeSymbol` is given directly and not its `OriginalDefinition` - which might create problems with Generic interfaces.